src/error.rs

```diff
@@ -1326,17 +1326,24 @@ impl ToDiagnostic<FileId> for TypecheckError {
                         format!("Maybe you forgot to put a `forall {}.` somewhere in the enclosing type ?", ident),
                     ])]
             ,
-            TypecheckError::TypeMismatch(expd, actual, span_opt) =>
+            TypecheckError::TypeMismatch(expd, actual, span_opt) => {
+                fn get_note(ty: &Types) -> String {
+                    if ty.0.is_flat_type() {
+                        String::from("the term")
+                    } else {
+                        String::from("a type expression")
+                    }
+                }
                 vec![
                     Diagnostic::error()
                         .with_message("incompatible types")
                         .with_labels(mk_expr_label(span_opt))
                         .with_notes(vec![
-                            format!("The type of the expression was expected to be `{}`", expd),
-                            format!("The type of the expression was inferred to be `{}`", actual),
+                            format!("Expected an expression typed by {} `{}`", get_note(expd), expd),
+                            format!("Inferred this expression to be typed by {} `{}`", get_note(actual), actual),
```

Comment:
A couple remarks:

- The new formulation (the beginning of the sentence) works nicely, but we use the formula "The type of the expression was expected to be" in a lot of places. I'm not too keen on having different error messages for no apparent reasons, so I think we have to change them all or to change none.
- "an expression typed by the term" doesn't sound right. We don't really have a standard definition for what is a type expression neither.

Here is a suggestion for a minimal change: keep the original structure of the error message, and only add something if one is a flat type with a variable inside and the other is a type variable:

```rust
                fn addendum(ty: &Types, other: &) -> &str {
                    match ty.0 {
                        AbsType::Flat(t) if !matches(t.as_ref(), Term::Var(_)) => " (contract)",
                        AbsType::Var(_) => " (type variable)",
                        _ => "",
                    }
                }
                vec![
                    Diagnostic::error()
                        .with_message("incompatible types")
                        .with_labels(mk_expr_label(span_opt))
                        .with_notes(vec![
                            format!("The type of the expression was expected to be `{}`{}", expd, addendum(expd)),
                            format!("The type of the expression was inferred to be `{}`{}", actual, addendum(actual)),
                            String::from("These types are not compatible"),
                        ])]
```

Reply:
Qre we sure vars are the only things which can cause confusion? could we have it with records?